scientific-computing-with-python/learn-special-methods-by-building-a-vector-space/vectors.py

```diff
@@ -0,0 +1,87 @@
+class Vector:
+    space = 'R3'
+
+    def __init__(self, x, y, z):
+        self.x = x
+        self.y = y
+        self.z = z
+        self.norm = (x**2 + y**2 + z**2)**0.5
+
+    def __str__(self):
+        return str((self.x, self.y, self.z))
+    
+    def __repr__(self):
+        return f'Vector({self.x}, {self.y}, {self.z})'
+
+    def __add__(self, other):        
+        x = self.x + other.x
+        y = self.y + other.y
+        z = self.z + other.z
+        v = Vector(x, y, z)
+        return v
+    
+    def __sub__(self, other):
+        x = self.x - other.x
+        y = self.y - other.y
+        z = self.z - other.z
+        v = Vector(x, y, z)
+        return v
+    
+    def __mul__(self, other):
+        # Scalar multiplication
+        if isinstance(other, int) or isinstance(other, float):
+            self.x *= other
+            self.y *= other
+            self.z *= other
+            self.norm *= other
+            return self
```

Comment:
I'd consider returning completely new `Vector` instance with multiplied attributes, instead of modifying self. Similarly like in other methods.